mautic/mautic
# Permission class DI

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

Each `Mautic[Plugin]\[bundle name]\Security\Permissions\[bundle name]Permissions` class was created on the fly without possibility to add dependencies to the class via constructor. This PR adds this possibility while keeping BC.

Here is how a permission class can be defined in a `app/bundles/[bundle name]/Config/config.php` array:
```php
    'services' =>
        'permissions' => [
            'sample.permissions' => [
                'class'     => \MauticPlugin\SampleBundle\Security\Permissions\SamplePermissions::class,
                'arguments' => [
                    '%mautic.parameters%',
                    'mautic.sample.model',
                    'translator',
                ],
            ],
        ],
    ],
```
It is still necessary to add `'%mautic.parameters%',` as a dependency as the `AbstractPermissions` class needs it to its constructor. So constructor in the `SamplePermissions` would look like this:
```php
    public function __construct(
        array $params,
        SampleModel $sampleModel,
        TranslatorInterface $translator
    )
    {
        parent::__construct($params);

        $this->sampleModel = $sampleModel;
        $this->translator = $translator;
    }
```
See an example how the AssetBundle creates the permission class with DI instead of the old way in commit `105d17c`.

#### Steps to test this PR:
1. Make sure that permission form still work and you can edit it.
2. Log in as a user with selected permissions and ensure that the user can view/create/edit/delete the entities he should be able and cannot view/create/edit/delete entities he shouldn't be able to. Test with the Asset bundle and one more bundle permissions.

#### List deprecations along with the new alternative:
1. It's recommended to create permission objects via DI. The old way of loading the classes with file system discovery is deprecated.
